### PR TITLE
Deduplicate code

### PR DESCRIPTION
**Pull request by mvirkkunen**
Implement the atomic operations that are missing on thumbv6 using critical sections to get rid of the duplication in atomic.rs and cm_mutex.rs. As an added bonus this also makes the critical sections even shorter than before.

Some documentation should still probably be copied over from cm_mutex.rs, such as the indication about which methods take a critical section and maybe there's other housekeeping stuff, but here's the general idea anyways. I've ran the tests on a PC and also made sure it compiles on thumbv6. Is there a way to actually test on thumbv6..?

### CODE REVIEW

File: core/src/bbbuffer.rs [hunk 1]
```diff
@@ -836,3 +836,53 @@ where
...
+            prev
+        })
+    }
+}
+
+#[cfg(not(feature = "thumbv6"))]
+mod atomic {
+    use core::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+
+    pub fn fetch_add(atomic: &AtomicUsize, val: usize, order: Ordering) -> usize {
```
Comment (jamesmunns) [inline core/src/bbbuffer.rs:877]:
```suggestion
    #[inline(always)]
    pub fn fetch_add(atomic: &AtomicUsize, val: usize, order: Ordering) -> usize {
```

File: core/src/bbbuffer.rs [hunk 2]
```diff
@@ -836,3 +836,53 @@ where
...
+        atomic.fetch_add(val, order)
+    }
+
+    pub fn fetch_sub(atomic: &AtomicUsize, val: usize, order: Ordering) -> usize {
```
Comment (jamesmunns) [inline core/src/bbbuffer.rs:881]:
```suggestion
    #[inline(always)]
    pub fn fetch_sub(atomic: &AtomicUsize, val: usize, order: Ordering) -> usize {
```

File: core/src/bbbuffer.rs [hunk 3]
```diff
@@ -836,3 +836,53 @@ where
...
+        atomic.fetch_sub(val, order)
+    }
+
+    pub fn swap(atomic: &AtomicBool, val: bool, order: Ordering) -> bool {
```
Comment (jamesmunns) [inline core/src/bbbuffer.rs:885]:
```suggestion
    #[inline(always)]
    pub fn swap(atomic: &AtomicBool, val: bool, order: Ordering) -> bool {
```

File: core/src/lib.rs [hunk 4]
```diff
@@ -91,17 +91,8 @@
...
-#[cfg(all(feature = "atomic", not(feature = "thumbv6")))]
-pub use atomic::*;
-
-#[cfg(feature = "thumbv6")]
-pub mod cm_mutex;
-
-#[cfg(all(feature = "thumbv6", not(feature = "atomic")))]
-pub use cm_mutex::*;
+mod bbbuffer;
+pub use bbbuffer::*;
```
Comment (jamesmunns) [inline core/src/lib.rs:95]: We probably need to also re-export each version under the `crate::atomic` and `crate::cm_mutex` namespaces to make this not a breaking change

Comment (mvirkkunen) [reply core/src/lib.rs:95]: I added some re-exports, would be nice if you could check if they're correct. Too bad the deprecated attribute doesn't work on modules, but I put one there anyways.

**Comment by mvirkkunen** [comment]
IMHO I would solve the problem of doctests not building with "thumbv6" by not running doctests with that feature at all. The code is exactly the same with that feature and without it, and you can't feasibly run the tests anyways. If it's OK I'd remove this line from `ci/test.sh`:

```cargo test --manifest-path core/Cargo.toml --no-default-features --features thumbv6```

Creating a separate thumbv6 test suite that can be run on actual hardware would be a good idea though.

**Pull request merged**